src/libosd/cl_cdm.c

```diff
@@ -0,0 +1,157 @@
+
+    size_t exp_payload_len = 1; //stall
+    assert(osd_packet_sizeconv_payload2data(exp_payload_len)
+           == pkg->data_size_words
+           && "CDM Protocol violation detected.");
+
+    bool stall = pkg->data.payload[0];
+  
+    ev->stall = stall;
+    return ev;
+}
+
+API_EXPORT
+osd_result osd_cl_cdm_handle_event(void *arg, struct osd_packet *pkg)
+{
+    assert(arg &&
+           "You need to give an event_handler_arg of type "
+           "struct osd_cdm_event_handler in osd_hostmod_new()");
+    assert(pkg);
+
+    struct osd_cdm_event_handler *handler = arg;
+
+    struct osd_cdm_event *ev = build_cdm_event(handler->cdm_desc, pkg);
+    osd_packet_free(&pkg);
+
+    handler->cb_fn(handler->cb_arg, handler->cdm_desc, ev);
+
+    free(ev);
+
+    return OSD_OK;
+}
+
+static bool is_cdm_module(struct osd_hostmod_ctx *hostmod_ctx,
+                          unsigned int cdm_di_addr)
+{
+    osd_result rv;
+
+    struct osd_module_desc desc;
+
+    rv = osd_hostmod_mod_describe(hostmod_ctx, cdm_di_addr, &desc);
+    if (OSD_FAILED(rv)) {
+        return false;
+    }
+
+    if (desc.vendor != OSD_MODULE_VENDOR_OSD ||
+        desc.type != OSD_MODULE_TYPE_STD_CDM || desc.version != 0) {
+        return false;
+    }
+
+    return true;
+}
+
+API_EXPORT
+osd_result osd_cl_cdm_get_desc(struct osd_hostmod_ctx *hostmod_ctx,
+                               unsigned int cdm_di_addr,
+                               struct osd_cdm_desc *cdm_desc)
+{
+    assert(hostmod_ctx);
+    assert(cdm_desc);
+
+    osd_result rv;
+
+    cdm_desc->di_addr = cdm_di_addr;
+
+    if (!is_cdm_module(hostmod_ctx, cdm_di_addr)) {
+        return OSD_ERROR_WRONG_MODULE;
+    }
+
+    uint16_t regvalue;
+    rv = osd_hostmod_reg_read(hostmod_ctx, &regvalue, cdm_di_addr,
+                              OSD_REG_CDM_CORE_CTRL, 16, 0);
+    if (OSD_FAILED(rv)) return rv;
+    cdm_desc->core_ctrl = regvalue;
+
+    rv = osd_hostmod_reg_read(hostmod_ctx, &regvalue, cdm_di_addr,
+                              OSD_REG_CDM_CORE_REG_UPPER, 16, 0);
+    if (OSD_FAILED(rv)) return rv;
+    cdm_desc->core_reg_upper = regvalue;
+
+    rv = osd_hostmod_reg_read(hostmod_ctx, &regvalue, cdm_di_addr,
+                              OSD_REG_CDM_CORE_DATA_WIDTH, 16, 0);
+    if (OSD_FAILED(rv)) return rv;
+    cdm_desc->core_data_width = regvalue;
+
+    return OSD_OK;
+}
+
+API_EXPORT
+/**
+ * Issue a read register request to the Core Debug Module (CDM)
+ */
+osd_result cl_cdm_cpureg_read(struct osd_hostmod_ctx *hostmod_ctx,
+                              unsigned int cdm_di_addr,
+                              void *reg_val, uint16_t reg_addr,
+                              int flags)
+{   
+    assert(hostmod_ctx);
+ 
+    osd_result rv;
+ 
+    rv = osd_hostmod_reg_read(hostmod_ctx, reg_val, cdm_di_addr, 
+                                    reg_addr, OSD_REG_CDM_CORE_DATA_WIDTH, 
+                                    flags);   
+    if (OSD_FAILED(rv)) return rv;
```

Comment:
That implementation is wrong in multiple ways.

1. You need to actually translate 'reg_addr' address to the real register address using the translation formula in the specification. You also need to make sure to set the CORE_REG_UPPER register to the right value if it's not already set correctly.

2. From my proposed prototype you stripped out the cdm desc parameter. This one is needed to get the current value of the the upper bits of the address.

3. OSD_REG_CDM_CORE_DATA_WIDTH is a define containing the register where the data width is stored. What you need here is actually the value inside this register.